### PR TITLE
WCOSS2 module and build updates

### PR DESCRIPTION
**Pull request by KateFriedman-NOAA**
In this PR are the following updates for WCOSS2 (being committed to the auth repo feature/ops-wcoss2 branch). This is a courtesy PR for comments ahead of submitting my fork branch changes into the auth repo branch. The full PR will happen when feature/ops-wcoss2 is merged into the operations branch. I will submit this PR if there are no objections or issues left to resolve next Tuesday. Thanks!

- add modulefiles for workflow owned codes (builds tested)
- add WCOSS2 stanzas in module-setup, machine-setup, link, and build scripts (tested)
- update the ufs-weather-model, GLDAS, and UPP to test versions for WCOSS2 (builds tested)

See a comment in related issue (https://github.com/NOAA-EMC/global-workflow/issues/399#issuecomment-921216364) for more details on these incoming changes to feature/ops-wcoss2 branch. No major changes will happen in this feature branch, only minimal updates to port to WCOSS2.

@EdwardSafford-NOAA please see the new WCOSS2 OznMon and RadMon build modulefiles.

The next PR will likely focus on the setup script updates. Rocoto will be installed on WCOSS2 soon so I can test the setup script produced xml when it's ready for use.

### CODE REVIEW

**Review by EdwardSafford-NOAA — COMMENTED** [review]
The OznMonBuild.wcoss2 and RadMonBuild.wcoss2 files can be deleted from the global_workflow.  All of the monitoring components are within the GSI and enter the workflow through the DA checkout these days.  My apologies for not noticing this earlier.  I'll take a look through the global_workflow to see if there are any more legacy *Mon files that can be removed.

File: modulefiles/module_base.wcoss2 [hunk 1]
```diff
@@ -0,0 +1,42 @@
+#%Module######################################################################
+##
+
+module load envvar/1.0
```
Comment (aerorahul) [inline modulefiles/module_base.wcoss2]: this is loaded in `modulefiles/module-setup.sh.inc`
```suggestion
```

File: modulefiles/module_base.wcoss2 [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+module load PrgEnv-intel/8.1.0
+module load craype/2.7.8
+module load intel/19.1.3.304
+module load cray-mpich/8.1.7
+module load cray-pals
```
Comment (aerorahul) [inline modulefiles/module_base.wcoss2:8]: curious, what is `cray-pals`?

File: util/sorc/compile_gfs_util_wcoss.sh [hunk 3]
```diff
@@ -10,7 +10,7 @@ LMOD_EXACT_MATCH=no
 source ../../sorc/machine-setup.sh > /dev/null 2>&1
 cwd=`pwd`
 
-if [ "$target" = "wcoss_dell_p3" ] || [ "$target" = "wcoss_cray" ] || [ "$target" = "hera" ] ; then
+if [ "$target" = "wcoss_dell_p3" ] || [ "$target" = "wcoss_cray" ] || [ "$target" = "hera" ] || [ "$target" = "wcoss2" ] ; then
```
Comment (aerorahul) [inline util/sorc/compile_gfs_util_wcoss.sh]: This line is checking for `hera` and the echo statement says `You are on WCOSS: $target`

File: util/sorc/mkgfsawps.fd/compile_mkgfsawps_wcoss.sh [hunk 4]
```diff
@@ -3,7 +3,7 @@ LMOD_EXACT_MATCH=no
 source ../../../sorc/machine-setup.sh > /dev/null 2>&1
 cwd=`pwd`
 
-if [ "$target" = "wcoss_dell_p3" ] || [ "$target" = "wcoss_cray" ] || [ "$target" = "hera" ] ; then
+if [ "$target" = "wcoss_dell_p3" ] || [ "$target" = "wcoss_cray" ] || [ "$target" = "hera" ] || [ "$target" = "wcoss2" ] ; then
```
Comment (aerorahul) [inline util/sorc/mkgfsawps.fd/compile_mkgfsawps_wcoss.sh]: same here.

**Comment by KateFriedman-NOAA** [comment]
> The OznMonBuild.wcoss2 and RadMonBuild.wcoss2 files can be deleted from the global_workflow. All of the monitoring components are within the GSI and enter the workflow through the DA checkout these days. My apologies for not noticing this earlier. I'll take a look through the global_workflow to see if there are any more legacy *Mon files that can be removed.

Thanks @EdwardSafford-NOAA ! I'll open a separate issue to remove those legacy files from the develop branch and will ping you in it for your inputs on that cleanup. 

Regarding removing those legacy files in the operations branch port to WCOSS2....@aerorahul @arunchawla-NOAA are we able to delete files in the transition to WCOSS2 or should we leave things as is if it doesn't need an update for WCOSS2? Thanks!

File: util/sorc/overgridid.fd/compile_overgridid_wcoss.sh [hunk 5]
```diff
@@ -10,7 +10,7 @@ LMOD_EXACT_MATCH=no
 source ../../../sorc/machine-setup.sh > /dev/null 2>&1
 cwd=`pwd`
 
-if [ "$target" = "wcoss_dell_p3" ] || [ "$target" = "wcoss_cray" ] || [ "$target" = "hera" ] ; then
+if [ "$target" = "wcoss_dell_p3" ] || [ "$target" = "wcoss_cray" ] || [ "$target" = "hera" ] || [ "$target" = "wcoss2" ] ; then
```
Comment (aerorahul) [inline util/sorc/overgridid.fd/compile_overgridid_wcoss.sh]: same here.

File: util/sorc/rdbfmsua.fd/compile_rdbfmsua_wcoss.sh [hunk 6]
```diff
@@ -10,7 +10,7 @@ LMOD_EXACT_MATCH=no
 source ../../../sorc/machine-setup.sh > /dev/null 2>&1
 cwd=`pwd`
 
-if [ "$target" = "wcoss_dell_p3" ] || [ "$target" = "wcoss_cray" ] || [ "$target" = "hera" ] ; then
+if [ "$target" = "wcoss_dell_p3" ] || [ "$target" = "wcoss_cray" ] || [ "$target" = "hera" ] || [ "$target" = "wcoss2" ] ; then
```
Comment (aerorahul) [inline util/sorc/rdbfmsua.fd/compile_rdbfmsua_wcoss.sh]: and here.

File: util/sorc/webtitle.fd/compile_webtitle_wcoss.sh [hunk 7]
```diff
@@ -10,7 +10,7 @@ LMOD_EXACT_MATCH=no
 source ../../../sorc/machine-setup.sh > /dev/null 2>&1
 cwd=`pwd`
 
-if [ "$target" = "wcoss_dell_p3" ] || [ "$target" = "wcoss_cray" ] || [ "$target" = "hera" ] ; then
+if [ "$target" = "wcoss_dell_p3" ] || [ "$target" = "wcoss_cray" ] || [ "$target" = "hera" ] || [ "$target" = "wcoss2" ] ; then
```
Comment (aerorahul) [inline util/sorc/webtitle.fd/compile_webtitle_wcoss.sh]: here as well.

**Review by aerorahul — CHANGES_REQUESTED** [review]
most changes look ok.
Some comments and changes requested.

**Comment by aerorahul** [comment]
> > The OznMonBuild.wcoss2 and RadMonBuild.wcoss2 files can be deleted from the global_workflow. All of the monitoring components are within the GSI and enter the workflow through the DA checkout these days. My apologies for not noticing this earlier. I'll take a look through the global_workflow to see if there are any more legacy *Mon files that can be removed.
> 
> Thanks @EdwardSafford-NOAA ! I'll open a separate issue to remove those legacy files from the develop branch and will ping you in it for your inputs on that cleanup.
> 
> Regarding removing those legacy files in the operations branch port to WCOSS2....@aerorahul @arunchawla-NOAA are we able to delete files in the transition to WCOSS2 or should we leave things as is if it doesn't need an update for WCOSS2? Thanks!

If they are not going to be used for building these monitoring utilities, then just don't add them in this PR.
There is no point in adding these files in this PR, not use them, and then remove them in an subsequent PR.

File: modulefiles/module_base.wcoss2 (hunk 2 above)
Comment (KateFriedman-NOAA) [reply modulefiles/module_base.wcoss2:8]: Per Carolyn's presentation slide on "Running jobs":

```
• PBS Pro uses the Cray Parallel Application Launch Service (PALS)
to launch MPI jobs
• cray-pals is a module that needs to be loaded to access mpiexec
```

I haven't gotten to testing job submission yet, have added the module in anticipation of needing it at runtime.

```
module show cray-pals
----------------------------------------------------------------------------------------------
   /opt/cray/pe/lmod/modulefiles/core/cray-pals/1.0.12.lua:
----------------------------------------------------------------------------------------------
whatis("PALS - Parallel Application Launch Service")
prepend_path("PATH","/opt/cray/pe/pals/1.0.12/bin")
prepend_path("MANPATH","/opt/cray/pe/pals/1.0.12/man")
set_alias("mpirun","mpiexec")
```

File: modulefiles/module_base.wcoss2 (hunk 1 above)
Comment (KateFriedman-NOAA) [reply modulefiles/module_base.wcoss2]: ~~Removed from both module-setup.sh.inc and module-setup.csh.inc at 0e9541da17b5d451d7fd1c89e68d7b077caa9678. Thanks!~~

Comment (KateFriedman-NOAA) [reply modulefiles/module_base.wcoss2]: done, removed from module_base.wcoss2 and added back into module-setup.*.inc, thanks!

File: util/sorc/compile_gfs_util_wcoss.sh (hunk 3 above)
Comment (KateFriedman-NOAA) [reply util/sorc/compile_gfs_util_wcoss.sh]: Removed Hera reference

File: util/sorc/mkgfsawps.fd/compile_mkgfsawps_wcoss.sh (hunk 4 above)
Comment (KateFriedman-NOAA) [reply util/sorc/mkgfsawps.fd/compile_mkgfsawps_wcoss.sh]: Removed Hera reference

File: util/sorc/overgridid.fd/compile_overgridid_wcoss.sh (hunk 5 above)
Comment (KateFriedman-NOAA) [reply util/sorc/overgridid.fd/compile_overgridid_wcoss.sh]: Removed Hera reference

File: util/sorc/rdbfmsua.fd/compile_rdbfmsua_wcoss.sh (hunk 6 above)
Comment (KateFriedman-NOAA) [reply util/sorc/rdbfmsua.fd/compile_rdbfmsua_wcoss.sh]: Removed Hera reference

File: util/sorc/webtitle.fd/compile_webtitle_wcoss.sh (hunk 7 above)
Comment (KateFriedman-NOAA) [reply util/sorc/webtitle.fd/compile_webtitle_wcoss.sh]: Removed Hera reference

File: modulefiles/module_base.wcoss2 [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+module load esmf/8.1.0
+module load cfp/2.0.4
+setenv USE_CFP YES
+
+module load cray-python/3.8.5.1
+module load prod_envir/2.0.4
+module load gempak/7.14.0
+module load nco/4.7.9
```
Comment (aerorahul) [inline modulefiles/module_base.wcoss2]: `nco` depends on `netcdf`

**Comment by KateFriedman-NOAA** [comment]
@aerorahul Please recheck the modulefiles, I've reorganized them. Let me know if any are still in the wrong order, thanks!

@BoiVuong-NOAA I committed discussed updates to the util codes, please check I've made the updates correctly, thanks!

**Review by aerorahul — COMMENTED** [review]
just one more place of module loads

File: modulefiles/module_base.wcoss2 (hunk 8 above)
Comment (KateFriedman-NOAA) [reply modulefiles/module_base.wcoss2]: fixed, thanks!

**Review by BoiVuong-NOAA — APPROVED** [review]
All makefile files for util look good.

File: modulefiles/OznMonBuild.wcoss2 [hunk 9]
```diff
@@ -0,0 +1,18 @@
+#%Module################################################
```
Comment (aerorahul) [inline modulefiles/OznMonBuild.wcoss2]: Is this used in the build?
I thought @EdwardSafford-NOAA mentioned that gfsv16.1 does not use this modulefile, but instead uses the one from GSI.

**Review by aerorahul — APPROVED** [review]
modulefiles look ok to me.

File: modulefiles/OznMonBuild.wcoss2 (hunk 9 above)
Comment (KateFriedman-NOAA) [reply modulefiles/OznMonBuild.wcoss2]: Correct, it's not used per @EdwardSafford-NOAA's confirmation, so I removed the RadMon/OznMonBuild.wcoss2 I had initially added. Will use what comes from the GSI instead. I can retire the other platform copies of the RadMon/OznMonBuild files in the develop branch separately.

File: sorc/build_gfs_bufrsnd.sh [hunk 10]
```diff
@@ -21,6 +21,11 @@ if [ ! -d "../exec" ]; then
   mkdir ../exec
 fi
 
+if [ $target = "wcoss2" ]; then
+  export SIGIO_LIB4=$SIGIO_LIB
```
Comment (WalterKolczynski-NOAA) [inline sorc/build_gfs_bufrsnd.sh:25]: Aren't these usually set by the sigio module?

File: sorc/build_tropcy_NEMS.sh [hunk 11]
```diff
@@ -29,6 +29,11 @@ else
 fi
 
 source ../modulefiles/modulefile.storm_reloc_v6.0.0.$target
+if [ $target = "wcoss2" ]; then
+  SIGIO_LIB4=$SIGIO_LIB
```
Comment (WalterKolczynski-NOAA) [inline sorc/build_tropcy_NEMS.sh:33]: Ibid

File: sorc/checkout.sh [hunk 12]
```diff
@@ -24,9 +24,10 @@ echo $topdir
 echo fv3gfs checkout ...
 if [[ ! -d fv3gfs.fd ]] ; then
     rm -f ${topdir}/checkout-fv3gfs.log
-    git clone https://github.com/ufs-community/ufs-weather-model fv3gfs.fd >> ${topdir}/checkout-fv3gfs.log 2>&1
+    #git clone https://github.com/ufs-community/ufs-weather-model fv3gfs.fd >> ${topdir}/checkout-fv3gfs.log 2>&1
```
Comment (WalterKolczynski-NOAA) [inline sorc/checkout.sh]: Remove commented code

File: util/sorc/mkgfsawps.fd/compile_mkgfsawps_wcoss.sh [hunk 13]
```diff
@@ -3,7 +3,7 @@ LMOD_EXACT_MATCH=no
 source ../../../sorc/machine-setup.sh > /dev/null 2>&1
 cwd=`pwd`
 
-if [ "$target" = "wcoss_dell_p3" ] || [ "$target" = "wcoss_cray" ] || [ "$target" = "hera" ] ; then
+if [ "$target" = "wcoss_dell_p3" ] || [ "$target" = "wcoss_cray" ] || [ "$target" = "wcoss2" ] ; then
```
Comment (WalterKolczynski-NOAA) [inline util/sorc/mkgfsawps.fd/compile_mkgfsawps_wcoss.sh:6]: Is removing Hera as a target going to break anything there?

File: util/sorc/overgridid.fd/compile_overgridid_wcoss.sh [hunk 14]
```diff
@@ -10,7 +10,7 @@ LMOD_EXACT_MATCH=no
 source ../../../sorc/machine-setup.sh > /dev/null 2>&1
 cwd=`pwd`
 
-if [ "$target" = "wcoss_dell_p3" ] || [ "$target" = "wcoss_cray" ] || [ "$target" = "hera" ] ; then
+if [ "$target" = "wcoss_dell_p3" ] || [ "$target" = "wcoss_cray" ] || [ "$target" = "wcoss2" ] ; then
```
Comment (WalterKolczynski-NOAA) [inline util/sorc/overgridid.fd/compile_overgridid_wcoss.sh:13]: Ibid

File: util/sorc/rdbfmsua.fd/compile_rdbfmsua_wcoss.sh [hunk 15]
```diff
@@ -10,7 +10,7 @@ LMOD_EXACT_MATCH=no
 source ../../../sorc/machine-setup.sh > /dev/null 2>&1
 cwd=`pwd`
 
-if [ "$target" = "wcoss_dell_p3" ] || [ "$target" = "wcoss_cray" ] || [ "$target" = "hera" ] ; then
+if [ "$target" = "wcoss_dell_p3" ] || [ "$target" = "wcoss_cray" ] || [ "$target" = "wcoss2" ] ; then
```
Comment (WalterKolczynski-NOAA) [inline util/sorc/rdbfmsua.fd/compile_rdbfmsua_wcoss.sh:13]: Ibid

File: util/sorc/webtitle.fd/compile_webtitle_wcoss.sh [hunk 16]
```diff
@@ -10,7 +10,7 @@ LMOD_EXACT_MATCH=no
 source ../../../sorc/machine-setup.sh > /dev/null 2>&1
 cwd=`pwd`
 
-if [ "$target" = "wcoss_dell_p3" ] || [ "$target" = "wcoss_cray" ] || [ "$target" = "hera" ] ; then
+if [ "$target" = "wcoss_dell_p3" ] || [ "$target" = "wcoss_cray" ] || [ "$target" = "wcoss2" ] ; then
```
Comment (WalterKolczynski-NOAA) [inline util/sorc/webtitle.fd/compile_webtitle_wcoss.sh:13]: Ibid

**Review by WalterKolczynski-NOAA — CHANGES_REQUESTED** [review]

File: sorc/build_tropcy_NEMS.sh (hunk 11 above)
Comment (KateFriedman-NOAA) [reply sorc/build_tropcy_NEMS.sh:33]: Needed to support both stack (WCOSS2) and pre-stack (elsewhere).

File: sorc/build_gfs_bufrsnd.sh (hunk 10 above)
Comment (KateFriedman-NOAA) [reply sorc/build_gfs_bufrsnd.sh:25]: Needed to support both stack (WCOSS2) and pre-stack (elsewhere).

**Review by BoiVuong-NOAA — APPROVED** [review]
looks good !

File: util/sorc/webtitle.fd/compile_webtitle_wcoss.sh (hunk 16 above)
Comment (KateFriedman-NOAA) [reply util/sorc/webtitle.fd/compile_webtitle_wcoss.sh:13]: Added Hera back in.

File: util/sorc/rdbfmsua.fd/compile_rdbfmsua_wcoss.sh (hunk 15 above)
Comment (KateFriedman-NOAA) [reply util/sorc/rdbfmsua.fd/compile_rdbfmsua_wcoss.sh:13]: Added Hera back in.

File: util/sorc/overgridid.fd/compile_overgridid_wcoss.sh (hunk 14 above)
Comment (KateFriedman-NOAA) [reply util/sorc/overgridid.fd/compile_overgridid_wcoss.sh:13]: Added Hera back in.

File: util/sorc/mkgfsawps.fd/compile_mkgfsawps_wcoss.sh (hunk 13 above)
Comment (KateFriedman-NOAA) [reply util/sorc/mkgfsawps.fd/compile_mkgfsawps_wcoss.sh:6]: Added Hera back in.

File: sorc/checkout.sh (hunk 12 above)
Comment (KateFriedman-NOAA) [reply sorc/checkout.sh]: removed

**Review by BoiVuong-NOAA — APPROVED** [review]

**Comment by KateFriedman-NOAA** [comment]
Tested fresh clone and build_all of branch on Cactus, no issues, although I forgot to add WCOSS2 to the machine check for the gfs_util build in build_all.sh. Committed that @ ea5357a. Good to submit this.

**Review by BoiVuong-NOAA — APPROVED** [review]

**Comment by KateFriedman-NOAA** [comment]
@BoiVuong-NOAA Please sync merge your fork copy of feature/ops-wcoss2 with the updated auth repo copy now. You should be able to use the "Fetch upstream" button. Let me know via email or Slack if you run into issues. Thanks!

**Comment by BoiVuong-NOAA** [comment]
Thank You, Kate.

**Pull request merged**